linera-service/src/persistent/file.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    io::{self, BufRead as _, Write as _},
+    path::Path,
+};
+
+use anyhow::Context as _;
+use fs4::FileExt as _;
+
+use super::Persistent;
+
+/// A guard that keeps an exclusive lock on a file.
+pub struct Lock(fs_err::File);
+
+impl Lock {
+    /// Acquires an exclusive lock on a provided `file`, returning a [`Lock`] which will
+    /// release the lock when dropped.
+    pub fn new(file: fs_err::File, path: &Path) -> anyhow::Result<Self> {
+        file.file().try_lock_exclusive().with_context(|| {
+            format!(
+                "Error getting write lock \"{}\". Please make sure the file exists \
+                 and that it is not in use by another process already.",
+                path.display()
+            )
+        })?;
+
+        Ok(Lock(file))
+    }
+}
+
+impl Drop for Lock {
+    fn drop(&mut self) {
+        if let Err(error) = self.0.file().unlock() {
+            tracing::warn!("Failed to unlock wallet file: {error}");
+        }
+    }
+}
+
+pub struct File<T> {
```

Comment:
nit: The name `File` is a bit counter-intuitive, thus being a value persisted on file. Maybe `OnFile<T>` is a step in the right direction?

Reply:
Another approach (which was highly debated before but perhaps ok here) is to use the module name for the distinction between different implementations. Something like `persistence::file_system::Value` implementing the trait `persistence::(Persistent)Value`.

Reply:
This is a wrapper over the type, so I read the name `persistent::File<Wallet>` as ‘a persistent file of/containing a wallet’.  The grammar of this name is hurt a bit with the rename to `persistence`, as well as `file_system` (a `persistence::file_system::Value<Wallet>` is not a persistence file system value containing a wallet).  I agree that `File` should never be used unqualified!

Reply:
So if you swap the storage medium, it will be called `persistent::Web<Wallet>` or `persistent::Network<Wallet>` ?

Reply:
Something like `persistent::LocalStorageEntry<Wallet>` or `persistent::S3Bucket<Wallet>`, yes (the name refers to the entry that stores the value, not the medium the entry lives on).